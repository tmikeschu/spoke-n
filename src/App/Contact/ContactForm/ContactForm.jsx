import React, { Component } from "react"
import APIService from "../../APIService/APIService"
import "../../App.css"

const apiService = new APIService("https://spoken-api.herokuapp.com")

export default class ContactForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      contact: {
        name: "",
        email: "",
        message: ""
      },
      success: false,
      failure: false,
      error: ""
    }
    this.emptyForm = this.state.contact

    this.handleChange = this.handleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(event) {
    this.setState({
      ...this.state,
      contact: {
        ...this.state.contact,
        [event.target.name]: event.target.value
      }
    })
  }

  messageData = () => ({
    contact: this.state.contact
  })

  async handleSubmit(event, service, messageData) {
    event.preventDefault()
    const response = await service.post("api/v1/contact", messageData)
    this.handleResponse(response)
    return response
  }

  handleResponse(response) {
    this.setState(this.stateFor(response))
    setTimeout(() => {
      this.setState({
        success: false,
        failure: false
      })
    }, 3000)
  }

  stateFor = (response) => {
    const { status, data } = response
    return (
      status === 201 && { success: true, contact: this.emptyForm }
    ) || (
      status === 400 && { failure: true, error: String(data) }
    )
  }

  render() {
    return (
      <article className="component-container contact-form">
        <form onSubmit={event => this.handleSubmit(event, apiService, this.messageData())}>
          <input 
            required type="text" name="name" value={this.state.contact.name} 
            onChange={this.handleChange} placeholder="Your Name"
          />

          <input 
            required type="text" name="email" value={this.state.contact.email} 
            onChange={this.handleChange} placeholder="Email Address"
          />

          <input type="hidden" name="_subject" value="Website contact" />

          <textarea 
            required name="message" value={this.state.contact.message} 
            onChange={this.handleChange} placeholder="What's up?"
          />

          <input type="submit" value="Send" />

          <p style={{ display: this.state.success ? "" : "none" }}>
            Message Sent!
          </p>
          <p style={{ display: this.state.failure ? "" : "none" }}>
            Message failure: { this.state.error }
          </p>
        </form>
        <p>
          *Please allow ample time for a response as we are busy bicycling 15,000 miles.
        </p>
      </article>
    )
  }
}

