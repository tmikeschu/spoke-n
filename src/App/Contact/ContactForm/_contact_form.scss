.contact-form {
  width: 100%;
  form {
    display: flex;
    flex-flow: column nowrap;
    width: 50%;
    margin: 0 auto;
    textarea, input {
      width: 100%;
      margin: 8px auto;
      font-size: 16px;
      padding: 16px;
      box-sizing: border-box;
    }
    input {
      height: 32px;
    }
    textarea {
      resize: none;
      height: 96px;
    }
    input {
      height: auto;
    }
  }
  p {
    text-align: center;
    margin: 8px auto;
  }

  @media (max-width: 640px) {
    width: 100%;
    form {
      width: 100%;
    }
    p {
      text-align: left;
    }
  }
}
