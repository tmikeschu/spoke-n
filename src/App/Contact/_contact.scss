@import "./ContactForm/contact_form.scss";

.Contact {
  @include contentContainer;
  display: flex;
  flex-direction: column;
  h3,
  section {
    margin-bottom: 32px;
  }
  @media (max-width: 640px) {
    height: 100%;
    section {
      p {
      }
    }
  }
}
