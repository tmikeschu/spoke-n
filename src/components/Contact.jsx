import React, { Component } from 'react';
import '../stylesheets/App.css';

export default class Contact extends Component {
  render() {
    return (
      <article className="contact">
        <h3>We would love to hear from you!</h3>
        <p>Email us at <a href="mailto:thespokentour@gmail.com">thespokentour@gmail.com</a></p>
        <p>Follow us on Instagram <a href="https://www.instagram.com/thespokentour/">@thespokentour</a></p>
      </article>
    );
  }
}

