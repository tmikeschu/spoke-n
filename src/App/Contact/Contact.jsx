import React from "react"
import ContactForm from "./ContactForm/ContactForm"

const Contact = () => (
  <article className="Contact">
    <h3>We would love to hear from you!</h3>
    <section>
      <p>
        Email us at{" "}
        <a
          href="mailto:thespokentour@gmail.com"
          target="_blank"
          rel="noopener noreferrer"
        >
          thespokentour@gmail.com
        </a>
      </p>
      <p>
        Follow us on Instagram{" "}
        <a
          href="https://www.instagram.com/thespokentour/"
          target="_blank"
          rel="noopener noreferrer"
        >
          @thespokentour
        </a>
      </p>
    </section>
    <ContactForm />
  </article>
)

export default Contact
