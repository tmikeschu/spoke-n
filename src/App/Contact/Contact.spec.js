import React from "react"
import { shallow } from "enzyme"
import Contact from "./Contact"
import ContactForm from "./ContactForm/ContactForm"

describe("<Contact />", () => {
  const contact = shallow(<Contact />)

  it("renders a contact page", () => {
    expect(contact).toBeTruthy()
  })

  it("renders the correct content", () => {
    expect(contact.equals(
      <article className="contact">
        <h3>We would love to hear from you!</h3>
        <section>
          <p>
            Email us at 
            <a href="mailto:thespokentour@gmail.com" target="_blank" rel="noopener noreferrer">
              thespokentour@gmail.com
            </a>
          </p>
          <p>
            Follow us on Instagram 
            <a href="https://www.instagram.com/thespokentour/" target="_blank" rel="noopener noreferrer">
              @thespokentour
            </a>
          </p>
        </section>
        <ContactForm />
      </article>
    )).toBeTruthy()
  })

  describe("snapshot", () => {
    it("is valid", () => {
      expect(contact.getNodes()).toMatchSnapshot()
    })
  }) 
})

