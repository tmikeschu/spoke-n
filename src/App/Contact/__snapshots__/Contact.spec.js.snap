// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Contact /> snapshot is valid 1`] = `
Array [
  <article
    className="contact"
>
    <h3>
        We would love to hear from you!
    </h3>
    <section>
        <p>
            Email us at
            <a
                href="mailto:thespokentour@gmail.com"
                rel="noopener noreferrer"
                target="_blank"
            >
                thespokentour@gmail.com
            </a>
        </p>
        <p>
            Follow us on Instagram
            <a
                href="https://www.instagram.com/thespokentour/"
                rel="noopener noreferrer"
                target="_blank"
            >
                @thespokentour
            </a>
        </p>
    </section>
    <ContactForm />
</article>,
]
`;
